File: java/PlanesAndroid/app/src/main/java/com/planes/android/TextButton.java
package com.planes.android;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.support.v7.widget.AppCompatButton;
import android.util.AttributeSet;
import android.view.ViewGroup;

public class TextButton extends AppCompatButton {

    public TextButton(Context context) {
        super(context);
        init();
    }

    public TextButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public TextButton(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    public void init() {
        m_Paint = new Paint();
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        PlanesVerticalLayout.PlanesVerticalLayoutParams lp = (PlanesVerticalLayout.PlanesVerticalLayoutParams) getLayoutParams();
        m_Text = lp.getText();
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        m_Paint.setColor(Color.BLUE);
        m_Paint.setTextSize(10);
        m_Paint.setTextAlign(Paint.Align.LEFT);
        int curTextSize = 10;

        int width = getWidth();
        int height = getHeight();

        int centerX = width / 2;
        int centerY = height / 2;

        int textWidth = 10;
        int textHeight = 10;

        while (textWidth < width / 2 && textHeight < height / 2) {
            m_Paint.setTextSize(curTextSize + 10);
            Rect bounds = new Rect();
            m_Paint.getTextBounds(m_Text, 0, m_Text.length(), bounds);
            textHeight = bounds.height();
            textWidth = bounds.width();
            curTextSize += 10;
        }

        m_Paint.setTextSize(curTextSize);

        canvas.drawText(m_Text, centerX - textWidth / 2, centerY + textHeight / 2, m_Paint);
    }


    private Paint m_Paint;
    private String m_Text;

}

File: kotlin/PlanesAndroid/app/src/main/java/com/planes/android/customviews/TextButton.kt
package com.planes.android.customviews

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatButton
import com.planes.android.customviews.CanvasPaintUtilities.Functions.computeOptimalTextSizeOneLine
import com.planes.android.customviews.CanvasPaintUtilities.Functions.drawButtonShadow
import com.planes.android.customviews.CanvasPaintUtilities.Functions.drawTextFitToSizeOneLine
import com.planes.android.customviews.CanvasPaintUtilities.Functions.measureHeightOneLineText
import com.planes.android.customviews.CanvasPaintUtilities.Functions.measureWidthOneLineText
import com.planes.android.game.common.PlanesVerticalLayoutParams

class TextButton : AppCompatButton, ViewWithText {
    private lateinit var m_Paint: Paint
    private lateinit var m_Text: String
    private var m_TextSize = 10
    private var m_MinWidth = 0
    private var m_MinHeight = 0

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init()
    }

    fun init() {
        m_Paint = Paint()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val lp = layoutParams as PlanesVerticalLayoutParams
        if (!this::m_Text.isInitialized) m_Text = lp.getText()
        val measuredHeight = measureHeightOneLineText(heightMeasureSpec, m_Paint, m_Text)
        val measuredWidth = measureWidthOneLineText(widthMeasureSpec, m_Paint, m_Text)
        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    public override fun onDraw(canvas: Canvas) {
        //super.onDraw(canvas);
        m_Paint.color = Color.GRAY
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), m_Paint)
        m_Paint.color = Color.BLUE
        drawTextFitToSizeOneLine(m_Text, m_TextSize, canvas, m_Paint, width, height)
        m_Paint.color = Color.BLACK
        drawButtonShadow(canvas, m_Paint, width, height)
    }

    override fun getOptimalTextSize(maxTextSize: Int, viewWidth: Int, viewHeight: Int): Int {
        return computeOptimalTextSizeOneLine(m_Text, m_Paint, viewWidth, viewHeight, maxTextSize)
    }

    override fun setTextSize(textSize: Int) {
        m_TextSize = textSize
        invalidate()
    }


}